### DISCUSSION TITLE
How to extract image

### DISCUSSION THREAD

**Discussion by pranjal-jaiswal** [jsvine/pdfplumber#496] [Q&A]
How to extract the images of the pdf?

**Comment by samkit-jain** [comment]
Hi @pranjal-jaiswal Appreciate your interest in the library. You can use the `.images` property to extract the images in a page of a PDF.

```python
import pdfplumber

pdf = pdfplumber.open("file.pdf")

for page in pdf.pages:
    for image in page.images:
        do_something(image)
```

**Reply by pranjal-jaiswal** [reply]
Hi @samkit-jain, Thanks for the prompt reply and help. But **.images**  give list of dictionary object with details of the image. I have attached a sample bellow. I want to save these images and process OCR on them. Please help me in this if you can. I know one method of cropping the image out of the page but I want a better solution. Thanks again for your help.

{'x0': Decimal('438.420'), 'y0': Decimal('104.640'), 'x1': Decimal('776.580'), 'y1': Decimal('507.360'), 'width': Decimal('338.160'), 'height': Decimal('402.720'), 'name': 'Im0', 'stream': <PDFStream(398): raw=48424, {'Subtype': /'Image', 'Length': 48423, 'Filter': /'DCTDecode', 'BitsPerComponent': 8, 'ColorSpace': <PDFObjRef:378>, 'Width': 500, 'Height': 595, 'Interpolate': True, 'Type': /'XObject'}>, 'srcsize': (Decimal('500'), Decimal('595')), 'imagemask': None, 'bits': 8, 'colorspace': [[/'ICCBased', <PDFStream(397): raw=2599, {'Length': 2598, 'Filter': /'FlateDecode', 'N': 3, 'Alternate': /'DeviceRGB'}>]], 'object_type': 'image', 'page_number': 1, 'top': Decimal('104.640'), 'bottom': Decimal('507.360'), 'doctop': Decimal('104.640')}

**Reply by jsvine** [reply]
Hi @pranjal-jaiswal, unfortunately `pdfplumber` does not currently provide a method for extracting the images embedded in a PDF. You might try working with the [pdfminer](https://github.com/pdfminer/pdfminer.six) object directly, via `pdf.doc`; see https://github.com/jsvine/pdfplumber/discussions/456#discussioncomment-947380 for details.

**Reply by samkit-jain** [reply]
You can use something similar to the following. 
```python
from pdfminer.image import ImageWriter
from pdfminer.layout import LTImage

imagewriter = ImageWriter("/path/to/output/directory/")

for p in pdf.pages:
    for obj in p.layout:
        if isinstance(obj, LTImage):
            imagewriter.export_image(obj)
```

A word of caution though that so far I have been unable to extract `LTImage` objects. Like @jsvine referenced, you can try using the `PDFDocument` object and see if you are able to extract the `LTImage` objects in the PDF. You can also use the CLI tool [pdfimages](https://www.xpdfreader.com/pdfimages-man.html) for the same.

**Reply by pranjal-jaiswal** [reply]
Thanks a lot @samkit-jain and @jsvine for your help.

**Reply by luojunhui1** [reply]
i still have this problem in 2023, is there any efficient or recommended methods for me to extract the images in PDF?  the advice of @samkit-jain enlightens me to check the code of pdfminer, however, i can't find the way to transfrom the dict like 
```{'x0': 90.0, 'y0': 390.7, 'x1': 491.1, 'y1': 613.55, 'width': 401.1, 'height': 222.84999999999997, 'name': 'Image54', 'stream': <PDFStream(54): raw=33238, {'Type': /'XObject', 'Subtype': /'Image', 'Width': 1234, 
'Height': 686, 'ColorSpace': /'DeviceRGB', 'BitsPerComponent': 8, 'Interpolate': False, 'SMask': <PDFObjRef:55>, 'Filter': /'FlateDecode', 'Length': 33238}>, 'srcsize': (1234, 686), 'imagemask': None, 'bits': 8, 'colorspace': [/'DeviceRGB'], 'object_type': 'image', 'page_number': 7, 'top': 228.37, 'bottom': 451.21999999999997, 'doctop': 5279.889999999999}
```
 to a `LTImage ` object, could you give me any advice, thanks a lot.

**Comment by LiBinNLP** [comment]
```
def extract_pdf_images(cur_page, image_dir):
    images = cur_page.images
    for idx, image in enumerate(images):
        bbox = [image['x0'], cur_page.cropbox[3]-image['y1'],  image['x1'], cur_page.cropbox[3]-image['y0']]
        img_page = cur_page.crop(bbox=bbox)
        img_obj = img_page.to_image(resolution=500)
        page_number = image['page_number']
        image_name_prefix = f'{page_number}-{idx + 1}'
        image_name = f'{image_name_prefix}' + ".png"
        image_path = f'{image_dir}/{image_name}'
        img_obj.save(image_path)
```

**Reply by DeoLeung** [reply]
it works but sometimes i get
```bash
ValueError: Bounding box [-1.0003357, -1.0308926999999812, 1225.2925873, 715.2450167000001] is not fully within parent page bounding box (0.0, 0.0, 1224.57, 714.331)
```

**Comment by samanth-jain** [comment]
```
import pdfplumber
import io
from PIL import Image
from IPython.display import display

pdf = pdfplumber.open("sample-ww2.pdf")
all_images = []
for page in pdf.pages:
    for image in page.images:
        all_images.append(image) 

for img in all_images:
    image_dict = img
    image_stream = image_dict['stream']
    image_data = image_stream.get_data()
    image_bytes = io.BytesIO(image_data)
    real_image = Image.open(image_bytes)

    display(real_image)
```

**Reply by BettenWhite** [reply]
UnidentifiedImageError: cannot identify image file <_io.BytesIO object at 0x744f2c41ca40>

**Reply by limdingwen** [reply]
Worked for me! `UnidentifiedImageError` could be due to a corrupted image file, or an image type not supported by your copy of PIL?

**Reply by martin0258** [reply]
I also got the [UnidentifiedImageError](https://github.com/jsvine/pdfplumber/discussions/496#discussioncomment-12524670), and I'm sure my image is not corrupted.

(NOTICE: The following code only works for uncompressed images)
I have modified the code as below, which worked on my side, hopefully it can help someone:

```python
from PIL import Image
# `img_obj` is an `Image` property from pfdplumber, https://github.com/jsvine/pdfplumber#image-properties
# E.g., the first image: img_obj = page.images[0] 
img_stream = img_obj["stream"]

# get image properties from stream
stream_attrs = img_stream.attrs
bits_per_color = img_obj["bits"]
colorspace = str(img_obj["colorspace"])
width, height = img_obj["srcsize"]
img_data = img_stream.get_data()

# determine color space
if "DeviceRGB" in colorspace:
    mode = "RGB"
elif "DeviceGray" in colorspace:
    mode = "L"
else:
    raise ValueError(f"Unsupported colorspace: {colorspace}")

# Most of the real-world PDF images are 8-bit pixels, supported by Pillow nicely
if bits_per_color != 8:
    # For other types, special logic is required (not implemented here), theoretically we can handle 1, 8, 32
    # Pillow offers 1-bit pixels, 8-bit pixels, 32-bit signed integer pixels, and 32-bit floating point pixels.
    raise ValueError("Only supported 8 bits img")

image = Image.frombuffer(mode, (width, height), img_data, "raw", mode, 0, 1)
image.save("output.png", format="PNG")
